core/domain/stats_services.py

```diff
@@ -421,15 +476,22 @@ def update_exp_issues_for_new_exp_version(
 
     playthrough_ids = list(itertools.chain.from_iterable(
         issue.playthrough_ids for issue in exp_issues.unresolved_issues))
-    playthrough_models = (
+    playthrough_models_with_none = (
         stats_models.PlaythroughModel.get_multi(playthrough_ids))
+    playthrough_models = []
 
-    for playthrough_model in playthrough_models:
+    for playthrough_model in playthrough_models_with_none:
+        # Ruling out the possibility of None for mypy type checking.
+        assert playthrough_model is not None
+        playthrough_models.append(playthrough_model)
         playthrough = get_playthrough_from_model(playthrough_model)
 
         if 'state_names' in playthrough.issue_customization_args:
             state_names = (
                 playthrough.issue_customization_args['state_names']['value'])
+            # Ruling out the possibility of any other type for mypy type
```

Comment:
Why can't this be done using the type of issue_customization_args? (if we make it a typed dict with named keys, and specify the type of issue_customization_args['state_names'])? Ditto below in several places.

Reply:
We cannot use TypedDict here because:
In issue_customization_args, we are providing an arbitrary number of keys with different types of values.
eg: here we are providing str and int 
https://github.com/oppia/oppia/blob/3c8d9fa916bdd40105a042915162bdccb8ec6a62/core/domain/stats_services_test.py#L1018

Here we are providing only one key `state_names` whose type is Dict[str, List[str]]
https://github.com/oppia/oppia/blob/3c8d9fa916bdd40105a042915162bdccb8ec6a62/core/domain/stats_services_test.py#L1068

Here, we are providing different key `time_spent_in_exp_in_msecs`
https://github.com/oppia/oppia/blob/3c8d9fa916bdd40105a042915162bdccb8ec6a62/core/domain/stats_domain_test.py#L749

Reply:
If I remember correctly, there are 3 different types of issues, and each of them has a specific set of customization args. See extensions/issues.

Btw, don't justify your arguments with code from the tests. The tests may be wrong. Justify your arguments with code from the main code since that is what gets run in production.

Reply:
If we do Typedict for each `issue`, then we have some of the following issues:

1. As we know, Typedict only works with static keys ( by this I mean fetching a value from a dictionary using a string literal as we are doing above `playthrough.issue_customization_args['state_names']` ) so if we use any string variable here then MyPy throws an error like
```python
class EarlyQuitIssueDict(TypedDict):
    """dash"""

    state_name: Dict[str, str]
    time_spent_in_exp_in_msecs: Dict[str, int]

early_quit_dict: EarlyQuitIssueDict

early_quit_dict['state_name']   # MYPY will work fine.

early_quit_dict[some_string_variable] #here MYPY will throw following errors.


TypedDict key must be a string literal; expected one of ('state_names')  [misc]
TypedDict key must be a string literal; expected one of ('state_name', 'time_spent_in_exp_in_msecs')  [misc]
```

and there are places in the codebase where we are doing this
https://github.com/oppia/oppia/blob/a2fb9adca12ee1949aa2e1940f1b58f2a15f4d22/core/domain/stats_services.py#L374
https://github.com/oppia/oppia/blob/a2fb9adca12ee1949aa2e1940f1b58f2a15f4d22/core/domain/stats_services.py#L375

To suppress the errors we have to use ignores.


2. Currently, we are using asserting only for value ( as we are doing above `playthrough.issue_customization_args['state_names']['value']` )
But if we define each issue and assign the union of all to the `issue_customization_args` attribute of Playthrough then it will also throw an error and to suppress the error we have to use assert. Like below:
```python
class CyclicStateTransitionsIssueDict(TypedDict):
    """dash"""

    state_names: Dict[str, List[str]]


class EarlyQuitIssueDict(TypedDict):
    """dash"""

    state_name: Dict[str, str]
    time_spent_in_exp_in_msecs: Dict[str, int]


class MultipleIncorrectSubmissionsDict(TypedDict):
    """dash"""

    state_name: Dict[str, str]
    num_times_answered_incorrectly: Dict[str, int]


AllowedIssuesDict = Union[
    CyclicStateTransitionsIssueDict,
    EarlyQuitIssueDict,
    MultipleIncorrectSubmissionsDict
]

class Playthrough:
    """Domain object representing a learner playthrough."""

    def __init__(
        self,
        issue_customization_args: AllowedIssuesDict,
    ):

        self.issue_customization_args = issue_customization_args

# Now considering the above line -- that was mentioned in this thread.
        if 'state_names' in playthrough.issue_customization_args:
            state_names = (
                playthrough.issue_customization_args['state_names']['value']) # Here MyPy throws an error, because issue_customization_args is of Union type (AllowedIssuesDict), so it will throw the following error.
```
error log:
```
core/domain/stats_services.py:494: error: TypedDict "EarlyQuitIssueDict" has no key 'state_names'  [misc]
core/domain/stats_services.py:494: note: Did you mean "state_name"?
core/domain/stats_services.py:494: error: TypedDict "MultipleIncorrectSubmissionsDict" has no key 'state_names'  [misc]
```
The cause of errors is that `issue_customization_args` is a union of `AllowedIssuesDict` and each dict that was mentioned in AllowedIssuesDict does not contain the same keys.
So, to suppress these errors we have to use asserts and make `issue_customization_args` of the appropriate TypedDict type before using it.

Reply:
I see -- thanks for the clear explanation. Of all the possible approaches we can take, which would you recommend and why?

Reply:
If we add one more named key in the typedicts mentioned above as `Literal`. We'll be able to narrow the type down without asserting. This key has to be populated in `get_playthrough_from_model`

```py
class CyclicStateTransitionsIssueDict(TypedDict):
    """dash"""
    type: Literal['CyclicStateTransitionsIssue']
    state_names: Dict[str, List[str]]


class EarlyQuitIssueDict(TypedDict):
    """dash"""
    type: Literal['EarlyQuitIssue']
    state_name: Dict[str, str]
    time_spent_in_exp_in_msecs: Dict[str, int]


class MultipleIncorrectSubmissionsDict(TypedDict):
    """dash"""
    type: Literal['MultipleIncorrectSubmissions']
    state_name: Dict[str, str]
    num_times_answered_incorrectly: Dict[str, int]
```

Now considering the above line
```py
if 'state_names' in playthrough.issue_customization_args:

# this can be replaced by
if playthrough.issue_customization_args['type'] == 'CyclicStateTransitionsIssueDict':
   reveal_type(playthrough.issue_customization_args) # The type should be CyclicStateTransitionsIssueDict
```

I was looking into narrowing the Union of typedicts based on the named keys in mypy, but turns out it is not implemented yet. The issue can be found [here](https://github.com/python/mypy/issues/11080). I found another issue linked with this. They have implemented narrowing Union of typedicts based on Literals which can be found [here](https://github.com/python/mypy/issues/7944). The approach I mentioned above is based on this.

I am not sure if this will throw more mypy errors

Reply:
I have tried @aasiffaizal 's approach and it seems to me that MYPY can only differentiate between the union of typedicts when all the typeddicts have `type` key. 

So, My first concern is that currently `playthrough.issue_customization_args` does not contain `type` key -- how do we populate this type key? ( I have tried to alter the issue_customization_args and type key but I'm not sure if that's a not going to break anything).

But yeah, That's a great approach and we should keep this in mind while implementing something in the future -- so that this kind of issues does not occur again.

Reply:
> which would you recommend and why?

I would recommend 1st one because in that way we have a subset of values that an issue_customization_args can take and we need to use a lesser number of assertions ( basically we need to use assert at the time only when we are providing value to some other function )

If we use the 2nd approach then we have to use assert and casts ( because assert will not work with typeddict ) every time when we access the `issue_customization_args` dict's value like: `playthrough.issue_customization_args['state_names']` because in 2nd approach `issue_customization_args` is Union of several typeddicts, and `state_names` is defined in only one TypedDict not in all so mypy throws error for all other typeddict and to suppress the error we have use assert and casts.

Reply:
@sahiljoster32 Every Playthrough object has an issue_type key, right? That will determine the format of the issue_customization_args.

So can we do this at the level of the Playthrough object instead?

Reply:
In this comment (https://github.com/oppia/oppia/pull/15870#discussion_r947754632) I mentioned all the problems if use typedict for each issue.

Now, Aasif's approach can only solve 2nd point's issue, where I mentioned that to narrow down the type from Union of typeddicts to a particular typedict we have to use assertion and casts. But as mentioned by Aasif if we use an extra `if clause` then we can narrow down the type easily -- which is true. 

Now, the problems that are still unsolved are:
1. As mentioned in this comment (https://github.com/oppia/oppia/pull/15870#discussion_r947754632), if we see 1st point there I mentioned about dynamically accessing the dict and in that case, Aasif's approach will not work and MYPY still floods the console. Because in Aasif's approach we already knew the `issue_customization_args`'s type and we are narrowing down the type using a hardcoded key ('type')
```python
if 'state_names' in playthrough.issue_customization_args:

# this can be replaced by
if playthrough.issue_customization_args['type'] == 'CyclicStateTransitionsIssueDict':
   reveal_type(playthrough.issue_customization_args) # The type should be CyclicStateTransitionsIssueDict
```
but in dynamically accessed dicts we don't know the specific type and we cannot use if clause there ( as far as I know ).

2. To differentiate between different TypedDicts we need at least one key which contains a string Literal, like in Aasif's comment we have `type` key (https://github.com/oppia/oppia/pull/15870#discussion_r949095903). -- now the problem here is that we don't have any key in `issue_customization_args` which contains only string literal. And if we add a new key -- then I'm not confident because then in that case typing comes to a runtime level ( ideally typing should be not executed at runtime ).

Reply:
> @sahiljoster32 Every Playthrough object has an issue_type key, right? That will fully determine the format of the issue_customization_args.
> 
> So can we do this at the level of the Playthrough object instead (i.e. one level higher)?

To narrow down the `issue_customization_args`'s Union type ( union of typeddicts ) to a particular typeddict, we need a key on `issue_customization_args`. So that we can do this `if playthrough.issue_customization_args['type'] == 'CyclicStateTransitionsIssueDict':` and Mypy will narrow down the type for us and it will not throw an error at runtime.

So, are your saying to define a new key on `issue_customization_args` in the Playthrough domain object?

Reply:
No, I am wondering if you can use an analogue of the issue_customization_args approach for the Playthrough domain object instead. That is, define three separate types of Playthrough domain object instead of three separate types of issue customization arg dicts, and then do `if playthrough.issue_type == ...` so that Mypy can narrow down the type for the playthrough domain object (which in turn constrains what its issue customization dict must look like).

Would an approach like that work? I don't see adding a key to the issue customization args dict as being feasible, it is too big a change just for the benefit of typing IMO.

Reply:
If you are mentioning to do something like below:
```python
class CyclicStateTransitionsIssueDict(TypedDict):
    """dash"""

    state_names: Dict[str, List[str]]


class MultipleIncorrectSubmissionsDict(TypedDict):
    """dash"""

    state_name: Dict[str, str]
    num_times_answered_incorrectly: Dict[str, int]


class EarlyQuitIssueDict(TypedDict):
    """dash"""

    state_name: Dict[str, str]
    time_spent_in_exp_in_msecs: Dict[str, int]

class Playthrough:
    """Domain object representing a learner playthrough."""

    def __init__(
        self,
        exp_id: str,
        exp_version: int,
        issue_type: str,
        issue_customization_args: Union[MultipleIncorrectSubmissionsDict, EarlyQuitIssueDict, CyclicStateTransitionsIssueDict],
        actions: List[LearnerAction]
    ): ...


class MultipleIncorrectSubmissionsPlaythrough(Playthrough):

    def __init__(
        self,
        exp_id: str,
        exp_version: int,
        issue_type: str,
        issue_customization_args: MultipleIncorrectSubmissionsDict,
        actions: List[LearnerAction]
    ) -> None:
        super().__init__(
            exp_id, exp_version, issue_type, issue_customization_args, actions
        )


class EarlyQuitIssuePlaythrough(Playthrough):

    def __init__(
        self,
        exp_id: str,
        exp_version: int,
        issue_type: str,
        issue_customization_args: EarlyQuitIssueDict,
        actions: List[LearnerAction]
    ) -> None:
        super().__init__(
            exp_id, exp_version, issue_type, issue_customization_args, actions
        )


class CyclicStateTransitionsIssuePlaythrough(Playthrough):

    def __init__(
        self,
        exp_id: str,
        exp_version: int,
        issue_type: str,
        issue_customization_args: CyclicStateTransitionsIssueDict,
        actions: List[LearnerAction]
    ) -> None:
        super().__init__(
            exp_id, exp_version, issue_type, issue_customization_args, actions
        )
```

Then at every step we have to use different Playthrough domain object -- taking an example of the lines mentioned in this thread:

```python
        if 'state_names' in playthrough.issue_customization_args:
            state_names = (
                playthrough.issue_customization_args['state_names']['value'])
```
We have to use CyclicStateTransitionsIssuePlaythrough instead of playthrough

```python
        if 'state_names' in CyclicStateTransitionsIssuePlaythrough.issue_customization_args:
            state_names = (
                playthrough.issue_customization_args['state_names']['value'])
```

Now, my question is how we return `CyclicStateTransitionsIssuePlaythrough` from `get_playthrough_from_model` explicitly because currently `get_playthrough_from_model` return Playthrough domain object.


And I'm not sure like how `if playthrough.issue_type == ...` will work because here `playthrough` is of domain object type not typeddict.

Reply:
Should we use generics as shown below? We wouldn't have to add an extra key, we could just initialize a `Playthrough` object based on `issue_type`

```py
class Playthrough(Generic[T]):
    def __init__(
        self,
        exp_id: str,
        exp_version: int,
        issue_type: str,
        issue_customization_args: T,
        actions: List[LearnerAction]
    ):
         # rest of the code
```

We'll need to initialize an instance as shown below based on `issue_type`

```py
Playthrough[EarlyQuitIssueDict](playthrough_model.exp_id, playthrough_model.exp_version,
        playthrough_model.issue_type,
        playthrough_model.issue_customization_args, actions))
```

Reply:
The problem with this approach is that it only works when we are initializing `Playthrough` object directly, I mean by `Playthrough` class.

But the problem in this thread is that we are getting `Playthrough` object from `get_playthrough_from_model` and I think this approach will not work there -- because in generics we have to provide type explicitly and in `get_playthrough_from_model` method we are just initializing object. ( Please correct me if I'm missing something ).

Reply:
In `get_playthrough_from_model`, we can initialize `Playthrough` object based on the `issue_type` in it right?
Either by the conditional statement or by dict of mapping of `issue_type` and the corresponding typedict if that works. Wouldn't that narrow down the type of `issue_customization_args` in the `Playthrough`  instance?
 
```py
if playthrough_model.issue_type == 'EarlyQuit':
    return Playthrough[EarlyQuitIssueDict](playthrough_model.exp_id, playthrough_model.exp_version,
        playthrough_model.issue_type,
        playthrough_model.issue_customization_args, actions))
```

Reply:
Places where we have to use ignores:
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L433
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L441
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L450
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L457
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L478
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L491
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L374
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L375

Reply:
For LearnerAction we have to define different set of TypedDicts:
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_domain.py#L1266
Due to which we have to use more ignores and casts. Like below:
https://github.com/oppia/oppia/blob/b1ec86cb2b3471530e74840dabb5755ea1b2a692/core/domain/stats_services.py#L457

Reply:
This last thing isn't a typed dict, is it? I thought for typed dicts we explicitly state what the keys are, and provide a type for each one.

For the comment before that, I think in general after you have something like "if 'state_name' in exp_issue.issue_customization_args" you can narrow the type of the customization args down to a union of TypedDict.

Reply:
As per the discussion, for this issue, a different PR is going to be raised, and to keep track of the track this issue I have created a TODO issue https://github.com/oppia/oppia/issues/15995